### PR TITLE
[Feature Request] New CBR resources supported

### LINKED ISSUE #1025: Feature Request: Add CBR resource support
Hi there,
I'd like to add CBR resource support on Terraform side.

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:

- CBR policys: 
  - Type
    - `backup`: customer can setting the execution times of backup tasks, the backup cycle, and retention rules.
    - `replication`: customer can setting the execution times of replication tasks, the replication cycle, and retention rules.
  - Backup cycle
    - `days`: the range from monday to sunday (`MO`, `TU`, `WE`, `TH`, `FR`, `SA`, `SU`).
    - `interval`: the interval backup time ranges from 1 day to 30 days.
    - `backup_times`: the time rules when the backup started, the value is UTC time on the hour (maximum count of rules is 24).
  - `max_backups`
  - `retention_duration_days`

- CBR vaults:
  - operation type supports `backup` and `replication` (except disk type).
  - customer can associte some resource (ECS, EVS, SFS turbo) to vault for corresponding backups.
  - vault can use policy rules to backup user data.
  - disk type support auto expand.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1025 

**Special notes for your reviewer**:

**Release note**:

```release-note
1. new CBR policy resource supported.
2. new CBR vault resource supported.
3. provide a consistent user experience with the console.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

CBR policy: weekly backup policy
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccCBRV3Policy_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccCBRV3Policy_basic -timeout 360m -parallel 4
=== RUN   TestAccCBRV3Policy_basic
=== PAUSE TestAccCBRV3Policy_basic
=== CONT  TestAccCBRV3Policy_basic
--- PASS: TestAccCBRV3Policy_basic (47.80s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       47.851s
```
CBR policy: daily replication policy
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccCBRV3Policy_replication'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccCBRV3Policy_replication -timeout 360m -parallel 4
=== RUN   TestAccCBRV3Policy_replication
=== PAUSE TestAccCBRV3Policy_replication
=== CONT  TestAccCBRV3Policy_replication
--- PASS: TestAccCBRV3Policy_replication (16.48s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       16.541s
```
Server type vault
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccCBRV3Vault_server'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccCBRV3Vault_server -timeout 360m -parallel 4
=== RUN   TestAccCBRV3Vault_serverBasic
=== PAUSE TestAccCBRV3Vault_serverBasic
=== RUN   TestAccCBRV3Vault_serverReplication
=== PAUSE TestAccCBRV3Vault_serverReplication
=== CONT  TestAccCBRV3Vault_serverBasic
=== CONT  TestAccCBRV3Vault_serverReplication
--- PASS: TestAccCBRV3Vault_serverReplication (17.46s)
--- PASS: TestAccCBRV3Vault_serverBasic (280.55s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       280.601s
```
Disk type vault
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccCBRV3Vault_volume'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccCBRV3Vault_volume -timeout 360m -parallel 4
=== RUN   TestAccCBRV3Vault_volumeBasic
=== PAUSE TestAccCBRV3Vault_volumeBasic
=== CONT  TestAccCBRV3Vault_volumeBasic
--- PASS: TestAccCBRV3Vault_volumeBasic (65.24s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       65.285s
```
SFS turbo vault
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccCBRV3Vault_turbo'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccCBRV3Vault_turbo -timeout 360m -parallel 4
=== RUN   TestAccCBRV3Vault_turboBasic
=== PAUSE TestAccCBRV3Vault_turboBasic
=== RUN   TestAccCBRV3Vault_turboReplication
=== PAUSE TestAccCBRV3Vault_turboReplication
=== CONT  TestAccCBRV3Vault_turboBasic
=== CONT  TestAccCBRV3Vault_turboReplication
--- PASS: TestAccCBRV3Vault_turboReplication (15.47s)
--- PASS: TestAccCBRV3Vault_turboBasic (502.16s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       502.202s
```

### CODE REVIEW

File: huaweicloud/resource_huaweicloud_cbr_policy.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+)
+
+func resourceCBRPolicyV3() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCBRPolicyV3Create,
+		Read:   resourceCBRPolicyV3Read,
+		Update: resourceCBRPolicyV3Update,
+		Delete: resourceCBRPolicyV3Delete,
+
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_cbr_policy.go:26]: can the CBR policy support to be imported ?

File: huaweicloud/resource_huaweicloud_cbr_policy_test.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		CheckDestroy: testAccCheckASV1PolicyDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testCBRV3Policy_basic(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCBRV3PolicyExists(resourceName, &asPolicy),
+					resource.TestCheckResourceAttr(resourceName, "name", rName),
+					resource.TestCheckResourceAttr(resourceName, "enabled", "true"),
+				),
+			},
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_cbr_policy_test.go:51]: the update steps should be added,

File: huaweicloud/resource_huaweicloud_cbr_policy.go (hunk 1 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_cbr_policy.go:26]: It is supported now.

File: huaweicloud/resource_huaweicloud_cbr_policy_test.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_cbr_policy_test.go:51]: thanks, the update steps and import steps are supported.

File: huaweicloud/resource_huaweicloud_cbr_policy.go [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+						},
+						"interval": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							ExactlyOneOf: []string{"trigger_pattern.0.days"},
+							ValidateFunc: validation.IntBetween(1, 30),
+						},
+						"days": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_cbr_policy.go:66]: suggest to add ValidateFunc

File: docs/resources/cbr_policy.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+* `destination_project_id` - (Optional, String) Specifies the ID of the replication destination project, which is
+  mandatory for cross-region replication.
+  Required if `operation_type` is `replication`.
+
+* `max_backups` - (Optional, Int) Specifies the maximum number of retained backups.
+  The value can be -1 or ranges from 2 to 99999.
+  This parameter and `retention_duration_days` are alternative.
+
+* `retention_duration_days` - (Optional, Int) Specifies the duration (in days) for retaining backups.
+  The value ranges from 0 to 99999.
```
Comment (ShiChangkuo) [inline docs/resources/cbr_policy.md]: the value range is different with the schema definition, please check this.

File: docs/resources/cbr_policy.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+* `destination_project_name` - (Optional, String) Specifies the name of the replication destination project, which is
+  mandatory for cross-region replication.
+  Required if `operation_type` is `replication`.
+
+* `destination_project_id` - (Optional, String) Specifies the ID of the replication destination project, which is
+  mandatory for cross-region replication.
+  Required if `operation_type` is `replication`.
+
+* `max_backups` - (Optional, Int) Specifies the maximum number of retained backups.
+  The value can be -1 or ranges from 2 to 99999.
```
Comment (ShiChangkuo) [inline docs/resources/cbr_policy.md]: does `-1` can work well ?

File: huaweicloud/resource_huaweicloud_cbr_policy.go [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+func resourceCBRPolicyV3Read(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*config.Config)
+	client, err := config.CbrV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("error creating Huaweicloud CBR client: %s", err)
+	}
+
+	cbrPolicy, err := policies.Get(client, d.Id()).Extract()
+	if err != nil {
+		if _, ok := err.(golangsdk.ErrDefault404); ok {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_cbr_policy.go]: please use **CheckDeleted**

File: huaweicloud/resource_huaweicloud_cbr_policy.go [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+		return &policyODCreate
+	}
+	policyODCreate.MaxBackups = maxBackups.(int)
+	policyODCreate.RetentionDurationDays = durationDays.(int)
+
+	return &policyODCreate
+}
+
+func makeSchedule(frequency, backupType, duration, time string) string {
+	timeSlice := strings.Split(time, ":")
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_cbr_policy.go:263]: we should check the length of timeSlice

File: huaweicloud/resource_huaweicloud_cbr_policy.go [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+	frequency := triggerPatternRaw[0].(map[string]interface{})["frequency"].(string)
+	var backupType, duration string
+	if frequency == "WEEKLY" {
+		backupType = "BYDAY"
+		duration = triggerPatternRaw[0].(map[string]interface{})["days"].(string)
+	} else {
+		backupType = "INTERVAL"
+		duration = strconv.Itoa(triggerPatternRaw[0].(map[string]interface{})["interval"].(int))
+	}
+	for i, v := range triggerPatternRaw[0].(map[string]interface{})["backup_times"].([]interface{}) {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_cbr_policy.go]: suggest to use a variable to store `triggerPatternRaw[0].(map[string]interface{})`

File: docs/resources/cbr_policy.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  It supports seven days a week (`MO`, `TU`, `WE`, `TH`, `FR`, `SA`, `SU`) and this parameter is separated by a
+  comma (,) without spaces, between date and date during the configuration.
+
+* `interval` - (Optional, Int) Specifies the interval of backup. Required if `frequency` is `DAILY`.
+  Valid value is range from `1` to `30`.
+
+* `backup_times` - (Required, List) Specifies the backup time. 
+  Automated backups will be triggered at the backup time.
+  The current time is in the UTC format (`hh:mm`).
+  The minutes in the list must be set to `00` and the hours cannot be repeated.
```
Comment (ShiChangkuo) [inline docs/resources/cbr_policy.md]: the attribute and import sections are missing

**Review by ShiChangkuo — CHANGES_REQUESTED** [review]

File: huaweicloud/resource_huaweicloud_cbr_policy.go (hunk 7 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_cbr_policy.go:263]: it's added and will check at validate function.

File: docs/resources/cbr_policy.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    backup_times = ["06:00"]
+  }
+}
+```
+
+### replication policy
+
+```hcl
+variable "policy_name" {}
+variable "project_name" {}
```
Comment (ShiChangkuo) [inline docs/resources/cbr_policy.md]: the variable is misleading.

Comment (Lance52259) [reply docs/resources/cbr_policy.md]: thanks. now, it use "dest_region_name" and "dest_project_id".

File: docs/resources/cbr_policy.md (hunk 5 above)
Comment (Lance52259) [reply docs/resources/cbr_policy.md]: the wrong document description has been modified.

File: docs/resources/cbr_policy.md (hunk 4 above)
Comment (Lance52259) [reply docs/resources/cbr_policy.md]: the wrong document description has been modified.

File: huaweicloud/resource_huaweicloud_cbr_policy.go (hunk 6 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_cbr_policy.go]: thanks. follow this method

File: huaweicloud/resource_huaweicloud_cbr_policy.go (hunk 8 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_cbr_policy.go]: already edited

File: docs/resources/cbr_vault.md [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+  * `include_volumes` - (Optional, List) Specifies the array of disk IDs which will be included in the backup.
+  This parameter and `exclude_volumes` are alternative.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - A resource ID in UUID format.
+
+* `allocated` - The allocated capacity of the vault, in MB.
```
Comment (ShiChangkuo) [inline docs/resources/cbr_vault.md]: please convert the unit to GB

File: docs/resources/cbr_vault.md [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+
+* `spec_code` - The specification code.
+
+* `status` - The vault status.
+
+* `storage` - The name of the bucket for the vault.
+
+* `used` - The used capacity, in MB.
```
Comment (ShiChangkuo) [inline docs/resources/cbr_vault.md]: ditto

File: huaweicloud/resource_huaweicloud_cbr_vault.go [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+	d.SetId("")
+
+	return nil
+}
+
+func resourceIncludeVolume(volumes []interface{}) []vaults.ResourceExtraInfoIncludeVolumes {
+	includeVolumes := make([]vaults.ResourceExtraInfoIncludeVolumes, len(volumes))
+	if len(volumes) == 0 {
+		return includeVolumes
+	}
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_cbr_vault.go]: the three line is unnecessary

File: huaweicloud/resource_huaweicloud_cbr_vault.go [hunk 14]
```diff
@@ -0,0 +1,431 @@
...
+		}
+	}
+	return includeVolumes
+}
+
+func resourceExcludeVolume(volumes []interface{}) []string {
+	includeVolumes := make([]string, len(volumes))
+	if len(volumes) == 0 {
+		return includeVolumes
+	}
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_cbr_vault.go]: ditto

File: huaweicloud/resource_huaweicloud_cbr_vault.go [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+		ConsistentLevel: d.Get("consistent_level").(string),
+		ProtectType:     d.Get("protection_type").(string),
+		Size:            d.Get("size").(int),
+	}
+
+	return billing
+}
+
+//Take out and save all the IDs nested in the list of resources map as a new string array.
+func changeSliceElementType(slice []interface{}) []string {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_cbr_vault.go]: please define the function with a better name

File: huaweicloud/resource_huaweicloud_cbr_vault.go [hunk 16]
```diff
@@ -0,0 +1,431 @@
...
+	}
+	return result
+}
+
+func updateResources(d *schema.ResourceData, client *golangsdk.ServiceClient) error {
+	oldRaws, newRaws := d.GetChange("resources")
+	resType := d.Get("type").(string)
+
+	//Calculate the intersection of oldRaws and newRaws, and use the intersection to calculate the subtraction.
+	intersectionRaws := oldRaws.(*schema.Set).Intersection(newRaws.(*schema.Set))
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_cbr_vault.go]: maybe the line is unnecessary

File: docs/resources/cbr_vault.md (hunk 11 above)
Comment (Lance52259) [reply docs/resources/cbr_vault.md]: thanks, the unit has been updated to __GB__, and the type of `allocated` has been updated to float64 (the basis of TypeFloat).

File: docs/resources/cbr_vault.md (hunk 12 above)
Comment (Lance52259) [reply docs/resources/cbr_vault.md]: fixed, ditto

File: huaweicloud/resource_huaweicloud_cbr_vault.go (hunk 13 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_cbr_vault.go]: These extra codes has been removed.

File: huaweicloud/resource_huaweicloud_cbr_vault.go (hunk 14 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_cbr_vault.go]: Fixed

File: huaweicloud/resource_huaweicloud_cbr_vault.go (hunk 15 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_cbr_vault.go]: The function name has been updated to `getResourceIDs`.

File: huaweicloud/resource_huaweicloud_cbr_vault.go (hunk 16 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_cbr_vault.go]: These codes have been updated according to other resource codes.

**Review by ShiChangkuo — APPROVED** [review]

**Pull request merged**